File: cpp/triangle/triangle.cpp
#include "triangle.h"

namespace triangle {
  flavor kind(double s1, double s2, double s3) {
    std::array<double, 3> sides{s1, s2, s3};
    std::sort(sides.begin(), sides.end());

    if (sides[0] <= 0) { 
      throw std::domain_error("All sides should be greater than 0.");
    }

    if (sides[0] + sides[1] < sides[2]) { 
      throw std::domain_error("Check for triangle inequality.");
    }

    if (sides[0] == sides[1] && sides[0] == sides[2] && sides[1] == sides[2]) {
      return flavor::equilateral;
    } else if (sides[0] == sides[1] || sides[0] == sides[2] || sides[1] == sides[2]) {
      return flavor::isosceles;
    }

    return flavor::scalene;
  }
}

File: ruby/triangle/triangle.rb
class Triangle
  def initialize(sides)
    raise "Three sides to a triangle." unless sides.length == 3
    @a, @b, @c = sides.sort()
  end

  def equilateral?
    valid? and (@a == @b and @b == @c)
  end

  def isosceles?
    valid? and (@a == @b or @b == @c or @a == @c)
  end

  def scalene?
    valid? and (@a != @b and @b != @c and @a != @c)
  end

  private

  def triangle_inequality_satisfied?
    @a + @b >= @c and
      @b + @c >= @a and
      @a + @c >= @b
  end

  def nonzero?
    @a > 0 and @b > 0 and @c > 0 
  end

  def valid?
    triangle_inequality_satisfied? and nonzero?
  end
end
